### PR TITLE
[geometry] Add RenderEngine support for shared_ptr

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Towards #5842 and #21968.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)

### CODE REVIEW

**Review by jwnimmer-tri — COMMENTED** [review]
+@SeanCurtis-TRI for feature review of the `//geometry/...` code, please.
+@rpoyner-tri for feature review of the `//pydrake/...` code, please.

> Reviewed 10 of 11 files at r1, 10 of 10 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-:-OFslNSG4ncZgJm5Bbcg:b3figmz)* status: 1 unresolved discussion, LGTM missing from assignees SeanCurtis-TRI(platform),rpoyner-tri(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFslShp36jdP1ZHunez:-OFslShp36jdP1ZHunf-:bpzgbxb) (no related file):*
Working

Needs (separate) platform reviewer assigned.

**Review by SeanCurtis-TRI — COMMENTED** [review]
`//geometry/...` reviewed -- I have some open questions.

> Reviewed 3 of 11 files at r1, 3 of 10 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-:-OFxD-rKEUUPL2A1bwf_:bte446h)* status: 5 unresolved discussions, LGTM missing from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OFx65qJ6fgKRxTSPLkQ:b-ti4utc) (no related file):*
Is it the case that if a user instantiates a `RenderEngine` in python and passes it to `SceneGraph` that the engine instance *must* be copied? I.e., the instance they had in hand was only the seed for what exists in `SceneGraph`, but is not a handle *to* it. That seems slightly unpythonic. Is it obvious to the python user that their instance got copied and operations on the in-hand instance will have no bearing on the in-SceneGraph instance? What if we allowed `SceneGraph` to take a `shared_ptr<RenderEngine>`? Would that lead to a more expected pythonic experience?

Admittedly, hanging onto the `RenderEngine` after creation and messing with it is a bit of an anti-pattern. So, maybe it's not worrying about.
___
*[`geometry/geometry_state.h` line 157 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx4wA59LK7DcPsNetf:-OFx4wA59LK7DcPsNetg:bcvr5su) ([raw file](https://github.com/RobotLocomotion/drake/blob/9416731c5267191273e2caa8ad8134ed1cbd9a15/geometry/geometry_state.h#L157)):*
> ```objc
>   }
>   DeepCopySharedPtr(DeepCopySharedPtr&& other) noexcept {
>     std::swap(value_, other.value_);
> ```

nit: I would've expected move semantics more akin to that of shared pointer. `shared_ptr` would leave `other` with a `null` value. Same for move assignment. Is there an explicit reason for this counter-intuitive implementation?
___
*[`geometry/test/geometry_state_test.cc` line 4958 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFxC3YI7QsJ62IUN5Jh:-OFxC3YI7QsJ62IUN5Ji:b-2o7ks) ([raw file](https://github.com/RobotLocomotion/drake/blob/9416731c5267191273e2caa8ad8134ed1cbd9a15/geometry/test/geometry_state_test.cc#L4958)):*
> ```cpp
> TEST_F(GeometryStateNoRendererTest, PerceptionRoleWithoutRenderer) {
>   const InternalGeometry& geometry = *gs_tester_.GetGeometry(geometries_[0]);
>   ASSERT_EQ(gs_tester_.RendererCount(), 0u);
> ```

nit: the `u` is now quite the useless appendage.
___
*[`geometry/scene_graph.h` line 702 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx5q8A9Klos3UzV3nB:-OFx5q8A9Klos3UzV3nC:b-sjxcc6) ([raw file](https://github.com/RobotLocomotion/drake/blob/9416731c5267191273e2caa8ad8134ed1cbd9a15/geometry/scene_graph.h#L702)):*
> ```objc
>   //@{
> 
>   /** Adds a new render engine to this %SceneGraph.
> ```

BTW The overload is documented as the "non-copying variant". Is it worth being explicit that the engine gets copied in this main documentation?

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 2 unresolved discussions, LGTM missing from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform)

___
*[`geometry/geometry_state.h` line 157 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx4wA59LK7DcPsNetf:-OFxiXhS1XH6ZurI1549:b-89cq5y) ([raw file](https://github.com/RobotLocomotion/drake/blob/9416731c5267191273e2caa8ad8134ed1cbd9a15/geometry/geometry_state.h#L157)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

nit: I would've expected move semantics more akin to that of shared pointer. `shared_ptr` would leave `other` with a `null` value. Same for move assignment. Is there an explicit reason for this counter-intuitive implementation?
</blockquote></details>

Derp.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 2 unresolved discussions, LGTM missing from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OFyRT9I4lGfrbX6XtFD:b-knrbcl) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Is it the case that if a user instantiates a `RenderEngine` in python and passes it to `SceneGraph` that the engine instance *must* be copied? I.e., the instance they had in hand was only the seed for what exists in `SceneGraph`, but is not a handle *to* it. That seems slightly unpythonic. Is it obvious to the python user that their instance got copied and operations on the in-hand instance will have no bearing on the in-SceneGraph instance? What if we allowed `SceneGraph` to take a `shared_ptr<RenderEngine>`? Would that lead to a more expected pythonic experience?

Admittedly, hanging onto the `RenderEngine` after creation and messing with it is a bit of an anti-pattern. So, maybe it's not worrying about.
</blockquote></details>

I guess I'd say that anyway their engine will be copied at context-creation time, so having it retain reference equality / identity while living in SceneGraph as a kind if "default value" but not while part of the Context didn't seem like an important win.  The big picture I'm aiming for "if its destination is eventually the context, then you never get to keep a pointer".  Things in the context have value semantics, not identity semantics.  (This stands in contract with our recent pointer and memory tweaks for System subclasses, which do _not_ have value semantics.)

If the user wants to keep tabs on all instances of their custom engines, their `__init__` and/or `__deepcopy__` methods have the power to register their engines with a global pool (e.g., for debugging / introspection).

I do think it's a _bug_ for a user to try to mutate their engine using the retained pointer after adding it to a scene graph.  There might be a case for granting them const access, but mutable access into internal geometry state smells wrong, and in Python we can't really enforce constness so it seems like a bad idea to allow them to keep a pointer to the engine they gave us.

I suppose the best reason to avoid the extra copy would be for performance, but I'd hope an empty engine would be relatively cheap to copy?  The argument passed to `AddRenderEngine` should nominally always be empty (no geometry added yet).

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 2 unresolved discussions, LGTM missing from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OG0y8D1Cg-8JwYduj4u:b-7mblhj) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess I'd say that anyway their engine will be copied at context-creation time, so having it retain reference equality / identity while living in SceneGraph as a kind if "default value" but not while part of the Context didn't seem like an important win.  The big picture I'm aiming for "if its destination is eventually the context, then you never get to keep a pointer".  Things in the context have value semantics, not identity semantics.  (This stands in contract with our recent pointer and memory tweaks for System subclasses, which do _not_ have value semantics.)

If the user wants to keep tabs on all instances of their custom engines, their `__init__` and/or `__deepcopy__` methods have the power to register their engines with a global pool (e.g., for debugging / introspection).

I do think it's a _bug_ for a user to try to mutate their engine using the retained pointer after adding it to a scene graph.  There might be a case for granting them const access, but mutable access into internal geometry state smells wrong, and in Python we can't really enforce constness so it seems like a bad idea to allow them to keep a pointer to the engine they gave us.

I suppose the best reason to avoid the extra copy would be for performance, but I'd hope an empty engine would be relatively cheap to copy?  The argument passed to `AddRenderEngine` should nominally always be empty (no geometry added yet).
</blockquote></details>

But anyway, you're right that this is one of the key questions to decide in this pull request, so I'm open to more discussion with you and always want to hear what Rico things.  I'd rather take our time and get this right.

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM: on the `//geometry/...` stuff.

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-:-OG1Ev8Z9iYqCOYf1qVA:b-73vyuu)* status: 2 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OG1CIsv9zSiUZqe6h0j:b-muzyvi) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

But anyway, you're right that this is one of the key questions to decide in this pull request, so I'm open to more discussion with you and always want to hear what Rico things.  I'd rather take our time and get this right.
</blockquote></details>

In some ways, I wonder if life wouldn't have been better if the `SceneGraph` API had simply taken the `RenderEngine` factory function instead of the instance.  :)

I believe your hypothesis is correct that copying an empty engine should be cheap. The only cost is the infrastructure. `RenderEngineVtk` recreates all of the VTK pipeline artifacts. `RenderEngineGl`....shares the context?  Something like that. So, performance at engine registration is probably not the end of the world.

That said, changing the `SceneGraph` api from `const RenderEngine& renderer` to `std::shared_ptr<RenderEngine> renderer` seems like an unfettered win. It completely eliminates the copy, regardless and, generally, I think we've taught our users that all configuration (geometry, etc.) should pass through `SceneGraph`. So, the idea that they have a handle on an engine stored in the model would only be used in the most self-flagellating circumstances. I like the idea of not having to ask the question, "is the copy expensive", at all.

tl;dr

Assuming there are no hidden traps in upstream library resource allocation, copying empty `RenderEngine` should be an acceptable mechanism to ensure user's can't even *try* to mutate `SceneGraph` internals directly.

At the same time, I'm ok with simply telling them not to do that in favor of the benefits.

I'm good either way.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 2 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OG1FOYo8QQCGl81Ltld:bay45xd) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

In some ways, I wonder if life wouldn't have been better if the `SceneGraph` API had simply taken the `RenderEngine` factory function instead of the instance.  :)

I believe your hypothesis is correct that copying an empty engine should be cheap. The only cost is the infrastructure. `RenderEngineVtk` recreates all of the VTK pipeline artifacts. `RenderEngineGl`....shares the context?  Something like that. So, performance at engine registration is probably not the end of the world.

That said, changing the `SceneGraph` api from `const RenderEngine& renderer` to `std::shared_ptr<RenderEngine> renderer` seems like an unfettered win. It completely eliminates the copy, regardless and, generally, I think we've taught our users that all configuration (geometry, etc.) should pass through `SceneGraph`. So, the idea that they have a handle on an engine stored in the model would only be used in the most self-flagellating circumstances. I like the idea of not having to ask the question, "is the copy expensive", at all.

tl;dr

Assuming there are no hidden traps in upstream library resource allocation, copying empty `RenderEngine` should be an acceptable mechanism to ensure user's can't even *try* to mutate `SceneGraph` internals directly.

At the same time, I'm ok with simply telling them not to do that in favor of the benefits.

I'm good either way.
</blockquote></details>

My main worry with taking a `shared_ptr<RenderEngine>` in the C++ API was that we'd need to write the API documentation "no really, you must stop using this engine now, please pinky swear" which is somewhat at odds with what `shared_ptr<>` usually denotes.  Having the "stop using it" be absolutely enforced by cloning seemed safest and more natural.

I'll wait for Rico to weigh in before changing anything.

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:  pydrake stuff.

> Reviewed 3 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-:-OG2HExR9s3YFOD0DDif:b-8lcdjt)* status: 3 unresolved discussions

___
*[`bindings/pydrake/geometry/geometry_py_render.cc` line 82 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OG2DRiD4p3CWMXg2PYx:-OG2DRiD4p3CWMXg2PYy:bw3hp5x) ([raw file](https://github.com/RobotLocomotion/drake/blob/ebef3bc8c8acb13f350ccb327e8a5e27f603c91c/bindings/pydrake/geometry/geometry_py_render.cc#L82)):*
> ```cpp
>     // can post-process the return value.
>     auto make_python_deepcopy = [&]() -> py::object {
>       PYBIND11_OVERLOAD_INT(py::object, Base, "DoClone");
> ```

nit It's probably worth noting to the casual reader that control will fall through this  macro if there is no suitable DoClone method. That wasn't obvious to me until I went and read the macro.

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 1 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 3 unresolved discussions

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 3 unresolved discussions

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: 1 unresolved discussion, LGTM missing from assignee xuchenhan-tri(platform)

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFslShp36jdP1ZHunez:-OG2OWR73e9d5dVtNuew:bfqkacy) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

Needs (separate) platform reviewer assigned.
</blockquote></details>

+@xuchenhan-tri for platform review per schedule, please.
___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OFx65qJ6fgKRxTSPLkP:-OG2NmAdB1OtLWq16oMz:b-e9d5l4) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I buy that argument. It would put both c++ and python in the same pinky-swear space. In the absence of unique pointers (or simply making sure the pointer is never accessible in the first place), copying is the only protection we have left. I'm sold.
</blockquote></details>

Done (no changes).
___
*[`bindings/pydrake/geometry/geometry_py_render.cc` line 82 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-OG2DRiD4p3CWMXg2PYx:-OG2ONCdDfd5J2gplhjP:b-eug7ee) ([raw file](https://github.com/RobotLocomotion/drake/blob/ebef3bc8c8acb13f350ccb327e8a5e27f603c91c/bindings/pydrake/geometry/geometry_py_render.cc#L82)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

nit It's probably worth noting to the casual reader that control will fall through this  macro if there is no suitable DoClone method. That wasn't obvious to me until I went and read the macro.
</blockquote></details>

Indeed, my comment paragraph above primed you to expect an unconditional return.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: LGTM missing from assignee xuchenhan-tri(platform)

**Review by rpoyner-tri — COMMENTED** [review]
> Reviewed 2 of 11 files at r1, 2 of 10 files at r2, 2 of 2 files at r3, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353)* status: LGTM missing from assignee xuchenhan-tri(platform)

**Review by xuchenhan-tri — COMMENTED** [review]
Platform :lgtm: 

> Reviewed 2 of 11 files at r1, 5 of 10 files at r2, 2 of 2 files at r3, 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22353#-:-OG6ppLkCteRSJOqrHwk:b-cxwtpw)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees rpoyner-tri(platform),SeanCurtis-TRI(platform),xuchenhan-tri(platform)
